src/data/string/ustring_utf_8_decode_test.cpp

```diff
@@ -0,0 +1,174 @@
+
+TEST(ustring_utf_8_decode_test, utf_8_byte_sequences_are_empty)
+{  const char* p="";
+  EXPECT_THROW({chars_to_uchar(p, p + strlen(p));}, std::invalid_argument);
+}
+
+TEST(ustring_utf_8_decode_test, utf_8_single_byte_character_is_out_of_range)
+{
+  for (int c = 0x80; c < 0xBF; c++) {
+    std::string str(1u, static_cast<char>(c));
+    const char* p=str.c_str();
+    EXPECT_THROW({chars_to_uchar(p, p + str.size());}, std::invalid_argument)
+      << std::hex << c;
+  }
+  for (int c = 0xFE; c <= 0xFF; c++) {
+    std::string str(1u, static_cast<char>(c));
+    const char* p=str.c_str();
+    EXPECT_THROW({chars_to_uchar(p, p + str.size());}, std::invalid_argument)
+      << std::hex << c;
+  }
+}
+
+TEST(ustring_utf_8_decode_test, utf_8_byte_sequences_have_a_5_byte_sequence)
+{
+  { // min 5-byte sequence
+    const char str[] = {0xF8, 0x80, 0x80, 0x80, 0x80, 0x00};
+    const char* p=str;
+    EXPECT_THROW({chars_to_uchar(p, p + strlen(p));}, std::invalid_argument) << str;
+  }
+  { // max 5-byte sequence
+    const char str[] = {0xFB, 0xBF, 0xBF, 0xBF, 0xBF, 0x00};
+    const char* p=str;
+    EXPECT_THROW({chars_to_uchar(p, p + strlen(p));}, std::invalid_argument) << str;
+  }
+}
+
+TEST(ustring_utf_8_decode_test, utf_8_byte_sequences_have_a_6_byte_sequence)
+{
+  { // min 6-byte sequence
+    const char str[] = {0xFC, 0x80, 0x80, 0x80, 0x80, 0x80, 0x00};
+    const char* p=str;
+    EXPECT_THROW({chars_to_uchar(p, p + strlen(p));}, std::invalid_argument) << str;
+  }
+  { // max 6-byte sequence
+    const char str[] = {0xFD, 0xBF, 0xBF, 0xBF, 0xBF, 0xBF, 0x00};
+    const char* p=str;
+    EXPECT_THROW({chars_to_uchar(p, p + strlen(p));}, std::invalid_argument) << str;
+  }
+}
+
+TEST(ustring_utf_8_decode_test, utf_8_byte_sequences_end_with_imcomplete_byte_sequence)
+{
+  {
+    const char str[] = {0xE3, 0x00}; // 2 bytes are missing
+    const char* p=str;
+    EXPECT_THROW({chars_to_uchar(p, p + strlen(p));}, std::invalid_argument) << str;
+  }
+  {
+    const char str[] = {0xE3, 0x80, 0x00}; // 1 byte is missing
+    const char* p=str;
+    EXPECT_THROW({chars_to_uchar(p, p + strlen(p));}, std::invalid_argument) << str;
+  }
+}
+
+TEST(ustring_utf_8_decode_test, utf_8_byte_sequences_have_a_start_byte_not_enough_continuation_bytes)
+{
+  {
+    const char str[] = {0xE3, 0x80, // 1 byte is missing here
+                        0xE3, 0x80, 0x81, 0x00};
+    const char* p=str;
+    EXPECT_THROW({chars_to_uchar(p, p + strlen(p));}, std::invalid_argument) << str;
+  }
+}
+
+TEST(ustring_utf_8_decode_test, utf_8_byte_sequences_have_an_overlong_encoding)
+{
+  { // Example of an overlong ASCII character('/')
+    const char str[3][5] = {
+      {0xC0, 0xAF, 0x00},
+      {0xE0, 0x80, 0xAF, 0x00},
+      {0xF0, 0x80, 0x80, 0xAF, 0x00},
+    };
+    for (size_t i = 0; i < 3; i++) {
+      const char* p=str[i];
+      EXPECT_THROW({chars_to_uchar(p, p + strlen(p));}, std::invalid_argument) << str;
+    }
+  }
+  { // Maximum overlong sequences
```

Comment:
Other tests check minimum pattern first, then maximum.
Same as others, here, checking minimum and then maximum is better?